sources/agent/greynoise/greynoise.go

```diff
@@ -0,0 +1,187 @@
+package greynoise
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/projectdiscovery/uncover/sources"
+)
+
+const (
+	URL = "https://api.greynoise.io/v3/gnql"
+)
+
+type Agent struct{}
+
+func (agent *Agent) Name() string { return "greynoise" }
+
+func (agent *Agent) Query(session *sources.Session, query *sources.Query) (chan sources.Result, error) {
+	if session.Keys.GreyNoiseKey == "" {
+		return nil, errors.New("empty GreyNoise API key")
+	}
+
+	results := make(chan sources.Result)
+
+	go func() {
+		defer close(results)
+
+		scrollToken := ""
+		total := 0
+		done := false
+
+		pageSize := 1000
+		if query.Limit > 0 && query.Limit < pageSize {
+			pageSize = query.Limit
+		}
+
+		for !done {
+			req := &Request{
+				Query:      query.Query,
+				Size:       pageSize,
+				Scroll:     scrollToken,
+				ExcludeRaw: true,
+			}
+
+			apiResponse, err := agent.query(session, req)
+			if err != nil {
+				results <- sources.Result{Source: agent.Name(), Error: err}
+				return
+			}
+			if apiResponse == nil || len(apiResponse.Data) == 0 {
+				return
+			}
+
+			for _, item := range apiResponse.Data {
+				host := firstNonEmpty(
+					item.InternetScannerIntelligence.Metadata.Domain,
+					item.InternetScannerIntelligence.Metadata.RDNS,
+				)
+
```

Comment:
there’s no port extraction like other agents.

i can see that `example.json` scan data includes port info as well as JA3 & HASSH fingerprints tied to specific ports.